src/nvim/lua/vim.lua

```diff
@@ -1,2 +1,62 @@
 -- TODO(ZyX-I): Create compatibility layer.
-return {}
+--{{{1 package.path updater function
+-- Last inserted paths. Used to clear out items from package.[c]path when they
+-- are no longer in &runtimepath.
+local last_nvim_paths = {}
+local function _update_package_paths()
+  local new_last_nvim_paths = {}
+  local rtps = vim.api.nvim_list_runtime_paths()
+  for _, key in ipairs({'path', 'cpath'}) do
+    local orig_str = package[key] .. ';'
+    local pathtrails = {}
+    local pathtrails_ordered = {}
+    local orig = {}
+    local orig_paths = {}
+    -- Note: ignores trailing item without trailing `;`. Not using something
+    -- simpler in order to preserve empty items (stand for default path).
+    for s in orig_str:gmatch('[^;]*;') do
+      s = s:sub(1, -2)  -- Strip trailing semicolon
+      orig[#orig + 1] = s
+      orig_paths[s] = true
+      -- Find out path patterns. pathtrail should contain something like
+      -- /?.so, /?/init.lua, /?.lua. This allows not to bother determining what
+      -- correct suffixes are.
+      local pathtrail = s:match('[/\\][^/\\]*%?.*$')
+      if pathtrail and not pathtrails[pathtrail] then
+        pathtrails[pathtrail] = pathtrail:sub(1, 1)
+        pathtrails_ordered[#pathtrails_ordered + 1] = pathtrail
+      end
+    end
+    local new = {}
+    for _, rtp in ipairs(rtps) do
+      if not rtp:match(';') then
+        for _, pathtrail in pairs(pathtrails_ordered) do
+          local sep = pathtrails[pathtrail]
+          local new_path = rtp .. sep .. 'lua' .. pathtrail
+          new[#new + 1] = new_path
+          new_last_nvim_paths[new_path] = true
+        end
+      end
+    end
+    for _, orig_path in ipairs(orig) do
```

Comment:
Comment for this loop:

> Make sure parts of the path that are unrelated to &runtimepath are retained.

Also, using the boolean logic simplifier: http://electronics-course.com/boolean-algebra, it looks like

```
~((A * ~B) + B) -- This code
```

Can be reduced to

```
~A * ~B
```

Or:

```
if not last_nvim_paths[orig_path] and not new_last_nvim_paths[orig_path] then
  -- ...
end
```

Which seems to have a natural explanation:

> Retain paths from the original `package.[c]path` that weren't in `last_nvim_paths` nor in `new_last_nvim_paths`.